url_shortener/dto.py

```diff
@@ -4,6 +4,31 @@
 from typing import NamedTuple
 
 
-class User(NamedTuple):
-    """Sample User DTO, which is used to represent authenticated user in out system"""
+class User:
+    id: int
     email: str
+    password: str
+    username: str
+    token: str
+    urls: [str]
+
+    def __init__(self, email, username, password, id):
+        self.username = username
+        self.password = password
+        self.email = email
+        self.id = id
+        self.urls = []
+        self.generate_token()
+
+
+    def generate_token(self):
+        self.token = hash(self.username)
```

Comment:
User should be able to login from multiple locations with different tokens, so that they could logout only from one device, but not from another (for example, user logged in from their mobile phone and laptop, than they decided to logout from the phone, but keep the laptop session active).

I would rather not store login token in `User` entity, but keep a separate dictionary which maps tokens to users.

Also, it would be better if you generate random tokens instead of relying on `hash` function. For example, you can use [UUID](https://docs.python.org/3/library/uuid.html) module for that:

```python
import uuid

token = uuid.uuid4().hex
```

Reply:
Thank you! Although let's assume that we have 5 logged in users, and let's assume that for some reason all of them logged in in 3 different devices. Iterating through logged in users(there is a list) and then inside of this user, finding this UUID would be faster. Then iterating through all 15 elements.

userObject.token should be an List<str> and keep UUIDs 

What do you think?

Reply:
> 5 logged in users all of them logged in in 3 different devices. Iterating through logged in users(there is a list) and then inside of this user, finding this UUID would be faster. Then iterating through all 15 elements.

In first cases you will iterate in worst case through 3 * 5 = 15 items. This will take almost the same time as iterating through 15 elements. However, there are data structures that allow constant-time search for the element. You can use `dict` structure for that:

```python
tokens_to_users = {
    'abcd-abcd-abcdabcdabcd-0001': 'user1',
    'abcd-abcd-abcdabcdabcd-0002': 'user2',
    'abcd-abcd-abcdabcdabcd-0003': 'user1',
    ...    
}
user_id = tokens_to_user.get(token)
```

This will find a `user_id` in constant time by token and that will be faster than iterating through all elements.